### ISSUE TITLE
Crash (memory error) when starting GUI program "DOSMenu 2".

### ISSUE THREAD

**Issue by Banjo-Oz** [joncampbell123/dosbox-x#2404]
Loading DOSMENU2 causes a crash and memory error reports in the debug console ("corrupt MCB chain").

Sadly, the error report seems way beyond my understanding but since the same program loads fine without errors in "vanilla" DOSBox, I wanted to report this since I use this program a fair bit on Dosbian and other places.

[DOSMENU2.ZIP](https://github.com/joncampbell123/dosbox-x/files/6235843/DOSMENU2.ZIP)

**Comment by Wengier** [comment]
It appears that DOSMENU2 worked fine in DOSBox-X 0.82.10, but it does not work in later versions. Doing a git bisect shows that this is the first commit that does not work with the said program: https://github.com/joncampbell123/dosbox-x/commit/4bd97f3d08fc07456dbe83722eb5364ee894963e

The code has been further modified since then so that the VGA BIOS ROM size is bigger than 0x3000. By trying to override the VGA BIOS size using a different value via the "vga bios size override" option DOSMENU2 will at least show the graphical screen, although it won't show all items properly.

**Comment by joncampbell123** [comment]
It scribbles over the VGA BIOS, possibly because of lazy SVGA programming. If the VGA BIOS is small enough, it will then corrupt the DOSBox private area and the UMB blocks.

DOSBox SVN as far as I know makes the VGA BIOS 32KB by default, which is probably why you never see this issue there.

It runs if you set:

[dos]
umb=false

I've seen similar issues with "Majic show" in which the fancy optimized video code manages to reach just below segment A000h and corrupt 2-4 bytes in segment 9FFFh, triggering a corrupt MCB chain message because the link to the UMB is broken.

**Comment by joncampbell123** [comment]
Also, the whole reason for the "segment limit" checks is other classes of demoscene and gaming that uses bank switched SVGA modes, and uses either segment limit violations or page faults to automatically bank switch while treating it like a linear framebuffer. Prior to that addition, these demos always used the memory at A0000h like a linear framebuffer and would draw well into the VGA BIOS, UMBs, and private memory area.

**Comment by joncampbell123** [comment]
https://user-images.githubusercontent.com/6245486/113229343-a5fdac80-924b-11eb-8bc6-4d9a6a91b762.mp4

**Comment by Wengier** [comment]
@joncampbell123 Yes, it will work if the UMB memory is disabled (or with the setting "video vga bios size override=-12288" in the current code). Very interesting.

**Comment by Banjo-Oz** [comment]
The program also loads for me with umb=true and xms=false!
